### PR TITLE
Further optimize storage of icu locale data

### PR DESCRIPTION
**Pull request by benaadams**
Shrink disk usage from 35kB to 16kB; skip .cctor initialization by switching to `ReadOnlySpan<byte>` rather than `ushort[]` and `int[]`

* Fix LcidToLocaleName lookup for Lcids that have sort order (was overflowing and truncating the sort bits)
* Optimize disk size of icu ThreeLetterWindowsLanguageName from 5184 bytes to 2592 bytes (`string` to `ReadOnlySpan<byte>`)
* Optimize disk size of `NameIndexToNumericData` 30kb to 13.5kb; by not using an `int` for every column and only storing the needed bytes.

Previous startup use:

image

No .cctor allocations in new path

Contributes to #45129

/cc @marek-safar @eerhardt

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs [hunk 1]
```diff
@@ -686,3100 +605,3131 @@ static void GenerateData (string[] cultures, (string, string)[] lcids)
...
-            0x51 << 16 | 279 << 4 | 2,       // bo
-            0x52 << 16 | 392 << 4 | 2,       // cy
-            0x53 << 16 | 1819 << 4 | 2,      // km
-            0x54 << 16 | 1966 << 4 | 2,      // lo
-            0x55 << 16 | 2141 << 4 | 2,      // my
-            0x56 << 16 | 1511 << 4 | 2,      // gl
-            0x57 << 16 | 1839 << 4 | 3,      // kok
-            0x58 << 16 | 2098 << 4 | 3,      // mni
-            0x59 << 16 | 2563 << 4 | 2,      // sd
-            0x5a << 16 | 2
```
Comment (benaadams) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs:3713]: Change to deal with the sort order prefix; they are grouped into sections (same order as before); however the sort is used for the lookup e.g. LCID lookup of `de-de` vs `de-de_phoneb`

Comment (benaadams) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs:3713]: Change to deal with the sort order prefix; they are grouped into sections (same order as before); however the sort is used for the lookup e.g. LCID lookup of `de-de` vs `de-de_phoneb`

Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs]: `LCIDToLocaleName` method; GitHub doesn't seem to be highlighting the correct lines 😢

**Comment by benaadams** [comment]
Might be a bug :)

File: src/libraries/System.Globalization/tests/CultureInfo/CultureInfoAll.cs [hunk 2]
```diff
@@ -525,71 +525,482 @@ private int GetLocaleInfoAsInt(CultureInfo ci, uint lcType)
...
+        }                               
+                                        
+        [Theory]                        
         [MemberData(nameof(CultureInfo_TestData))]
-        public void LcidTest(string cultureName, int lcid, string specificCultureName, string threeLetterISOLanguageName, string threeLetterWindowsLanguageName, string alternativeCultureName, string consoleUICultureName)
-        {
-            _ = alternativeCultureName;
+        public void LcidTest(int lcid, string[] cultureNames, string specificCultureName, string threeLetterISOLanguageName, string threeLetterWindowsLanguageName, string alternativeCultureName, string consoleUICultureName)
+        {                               
+            _ = alternativeCultureName;
```
Comment (am11) [inline src/libraries/System.Globalization/tests/CultureInfo/CultureInfoAll.cs:937]: nit: trailing spaces.
(might make it easier to validate/review, if test enhancements are split in a separate PR)

**Comment by danmosemsft** [comment]
cc @lewing

File: src/libraries/System.Globalization/tests/CultureInfo/CultureInfoAll.cs (hunk 2 above)
Comment (benaadams) [reply src/libraries/System.Globalization/tests/CultureInfo/CultureInfoAll.cs]: Extending the test helped pick up a couple issues in the data; e.g.

image

File: src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs [hunk 3]
```diff
@@ -22,89 +22,8 @@ internal enum IcuLocaleDataParts
...
-                Debug.Assert(str.Item1 < Math.Pow (2, 12));
-                Debug.Assert(str.Item2 < Math.Pow (2, 4));
-
-                Console.WriteLine($"{entry.Item1} << 16 | {str.Item1} << 4 | {str.Item2},\t // {name}");
-            }
-            Console.WriteLine("};");
-        }
-        */
-
         private static ReadOnlySpan<byte> CultureNames => new byte[]
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs:27]: If https://github.com/dotnet/runtime/pull/45699 is some form is merged; could change it to an actual source generator from input files? 🤔

**Comment by tarekgh** [comment]
@benaadams is there any changes you need to do here? or @marek-safar comments can be addressed in other PR and we can merge this one?

**Comment by benaadams** [comment]
Is good from me; cna do it in follow up

**Comment by tarekgh** [comment]
ok, merging this one and any further comments can be addressed in follow up PR's.

**Review by tarekgh — APPROVED** [review]
